File: go/headgateprometheus/prometheus.go
// Package headgateprometheus exports Headgate runner telemetry as Prometheus metrics.
// The application owns the registry, scrape endpoint, and Prometheus server.
package headgateprometheus

import (
	"fmt"

	"github.com/mujhtech/headgate/go"
	"github.com/prometheus/client_golang/prometheus"
)

// Telemetry translates Headgate's exporter-neutral runtime events into Prometheus
// collectors. Its methods are safe for concurrent use.
type Telemetry struct {
	admitted           *prometheus.CounterVec
	rejected           *prometheus.CounterVec
	completed          *prometheus.CounterVec
	quarantined        prometheus.Counter
	evicted            *prometheus.CounterVec
	duration           *prometheus.HistogramVec
	workerUtilization  *prometheus.GaugeVec
	workerEmptyPoll    *prometheus.GaugeVec
	workerInflight     *prometheus.GaugeVec
	workerCapacity     *prometheus.GaugeVec
	workerMemory       *prometheus.GaugeVec
	workerMemoryLimit  *prometheus.GaugeVec
	workerRestartCount *prometheus.CounterVec
}

// New registers a telemetry adapter with an application-owned registry. It rejects a
// nil registerer instead of silently using Prometheus's process-global registry. If any
// name collides, New unregisters the collectors it added before returning the error.
func New(registerer prometheus.Registerer) (*Telemetry, error) {
	if registerer == nil {
		return nil, fmt.Errorf("headgateprometheus: registerer must not be nil")
	}

	t := newTelemetry()
	collectors := []prometheus.Collector{
		t.admitted, t.rejected, t.completed, t.quarantined, t.evicted, t.duration,
		t.workerUtilization, t.workerEmptyPoll, t.workerInflight, t.workerCapacity,
		t.workerMemory, t.workerMemoryLimit, t.workerRestartCount,
	}
	registered := make([]prometheus.Collector, 0, len(collectors))
	for _, collector := range collectors {
		if err := registerer.Register(collector); err != nil {
			for _, added := range registered {
				registerer.Unregister(added)
			}
			return nil, fmt.Errorf("headgateprometheus: register metrics: %w", err)
		}
		registered = append(registered, collector)
	}
	return t, nil
}

var _ headgate.Telemetry = (*Telemetry)(nil)

// OnEvent implements headgate.Telemetry. Unknown event types and invalid negative event
// counts are ignored, keeping a future or malformed telemetry event off the job path.
func (t *Telemetry) OnEvent(event headgate.Event) {
	switch event.Type {
	case "admitted":
		addCount(t.admitted.WithLabelValues(event.Queue), event.Count)
	case "rejected":
		addCount(t.rejected.WithLabelValues(event.Queue, event.Policy), event.Count)
	case "completed":
		t.completed.WithLabelValues(event.Kind).Inc()
	case "quarantined":
		t.quarantined.Inc()
	case "evicted":
		addCount(t.evicted.WithLabelValues(event.Queue), event.Count)
	case "job_span":
		if event.Duration >= 0 {
			t.duration.WithLabelValues(event.Queue, event.Kind, event.Outcome).
				Observe(event.Duration.Seconds())
		}
	case "worker_saturation":
		t.workerUtilization.WithLabelValues(event.Worker).Set(event.Utilization)
		t.workerEmptyPoll.WithLabelValues(event.Worker).Set(event.EmptyPollRatio)
		t.workerInflight.WithLabelValues(event.Worker).Set(float64(event.Inflight))
		t.workerCapacity.WithLabelValues(event.Worker).Set(float64(event.Capacity))
	case "worker_memory":
		t.workerMemory.WithLabelValues(event.Worker).Set(float64(event.MemoryBytes))
		t.workerMemoryLimit.WithLabelValues(event.Worker).Set(float64(event.MemoryLimitBytes))
		if event.RestartRequested {
			t.workerRestartCount.WithLabelValues(event.Worker).Inc()
		}
	}
}

func addCount(counter prometheus.Counter, count int) {
	if count > 0 {
		counter.Add(float64(count))
	}
}

func newTelemetry() *Telemetry {
	// Dashboard: sum by (queue) (rate(headgate_jobs_admitted_total[5m]))
	admitted := prometheus.NewCounterVec(prometheus.CounterOpts{
		Namespace: "headgate", Subsystem: "jobs", Name: "admitted_total",
		Help: "Number of jobs admitted for execution.",
	}, []string{"queue"})
	// Dashboard: sum by (queue, policy) (rate(headgate_jobs_rejected_total[5m]))
	rejected := prometheus.NewCounterVec(prometheus.CounterOpts{
		Namespace: "headgate", Subsystem: "jobs", Name: "rejected_total",
		Help: "Number of runtime-visible policy rejections.",
	}, []string{"queue", "policy"})
	// Dashboard: sum by (kind) (rate(headgate_jobs_completed_total[5m]))
	completed := prometheus.NewCounterVec(prometheus.CounterOpts{
		Namespace: "headgate", Subsystem: "jobs", Name: "completed_total",
		Help: "Number of fence-verified durable job completions.",
	}, []string{"kind"})
	// Alert: increase(headgate_jobs_quarantined_total[10m]) > 0
	quarantined := prometheus.NewCounter(prometheus.CounterOpts{
		Namespace: "headgate", Subsystem: "jobs", Name: "quarantined_total",
		Help: "Number of jobs moved to quarantine.",
	})
	// Dashboard: sum by (queue) (rate(headgate_jobs_evicted_total[5m]))
	evicted := prometheus.NewCounterVec(prometheus.CounterOpts{
		Namespace: "headgate", Subsystem: "jobs", Name: "evicted_total",
		Help: "Number of terminal jobs evicted by retention.",
	}, []string{"queue"})
	// Dashboard: histogram_quantile(0.99, sum by (le, queue, kind) (rate(headgate_job_attempt_duration_seconds_bucket[5m])))
	duration := prometheus.NewHistogramVec(prometheus.HistogramOpts{
		Namespace: "headgate", Subsystem: "job", Name: "attempt_duration_seconds",
		Help:    "Job attempt duration in seconds.",
		Buckets: []float64{0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10, 30, 60, 300},
	}, []string{"queue", "kind", "outcome"})
	// Dashboard: avg(headgate_worker_utilization_ratio)
	workerUtilization := prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Namespace: "headgate", Subsystem: "worker", Name: "utilization_ratio",
		Help: "Current worker inflight-to-capacity ratio.",
	}, []string{"worker"})
	// Dashboard: avg(headgate_worker_empty_poll_ratio)
	workerEmptyPoll := prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Namespace: "headgate", Subsystem: "worker", Name: "empty_poll_ratio",
		Help: "Current ratio of empty admission polls in the worker rolling window.",
	}, []string{"worker"})
	// Dashboard: sum(headgate_worker_inflight)
	workerInflight := prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Namespace: "headgate", Subsystem: "worker", Name: "inflight",
		Help: "Current number of in-flight jobs.",
	}, []string{"worker"})
	// Dashboard: sum(headgate_worker_capacity)
	workerCapacity := prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Namespace: "headgate", Subsystem: "worker", Name: "capacity",
		Help: "Configured worker execution capacity.",
	}, []string{"worker"})
	// Dashboard: headgate_worker_memory_bytes / headgate_worker_memory_limit_bytes
	workerMemory := prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Namespace: "headgate", Subsystem: "worker", Name: "memory_bytes",
		Help: "Last sampled worker process memory footprint in bytes.",
	}, []string{"worker"})
	workerMemoryLimit := prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Namespace: "headgate", Subsystem: "worker", Name: "memory_limit_bytes",
		Help: "Configured worker process memory limit in bytes.",
	}, []string{"worker"})
	// Alert: increase(headgate_worker_restart_requests_total[15m]) > 0
	workerRestartCount := prometheus.NewCounterVec(prometheus.CounterOpts{
		Namespace: "headgate", Subsystem: "worker", Name: "restart_requests_total",
		Help: "Number of memory-guard restart requests.",
	}, []string{"worker"})

	return &Telemetry{
		admitted: admitted, rejected: rejected, completed: completed,
		quarantined: quarantined, evicted: evicted, duration: duration,
		workerUtilization: workerUtilization, workerEmptyPoll: workerEmptyPoll,
		workerInflight: workerInflight, workerCapacity: workerCapacity,
		workerMemory: workerMemory, workerMemoryLimit: workerMemoryLimit,
		workerRestartCount: workerRestartCount,
	}
}

File: crates/headgate-prometheus/tests/prometheus.rs
use headgate_core::{Event, Telemetry as _};
use prometheus::{Encoder, Gauge, Opts, Registry, TextEncoder};

#[test]
fn exports_matching_lifecycle_duration_and_worker_metrics() {
    let registry = Registry::new();
    let telemetry = headgate_prometheus::Telemetry::new(&registry).expect("register metrics");

    telemetry.on_event(Event::Admitted {
        queue: "critical",
        count: 3,
    });
    telemetry.on_event(Event::Rejected {
        queue: "critical",
        policy: "rate_class",
        count: 2,
    });
    telemetry.on_event(Event::Completed {
        kind: "mail.send",
        ms: 250,
    });
    telemetry.on_event(Event::JobSpan {
        job_id: "must-not-be-a-label",
        kind: "mail.send",
        queue: "critical",
        attempt: 1,
        outcome: "success",
        started_at_ms: 1_700_000_000_000,
        ms: 250,
        trace: None,
    });
    telemetry.on_event(Event::WorkerSaturation {
        worker: "worker-a",
        inflight: 3,
        capacity: 8,
        utilization: 0.375,
        empty_poll_ratio: 0.25,
        polls: 4,
        empty_polls: 1,
    });
    telemetry.on_event(Event::WorkerMemory {
        worker: "worker-a",
        used_bytes: 90,
        limit_bytes: 100,
        restart_requested: true,
    });

    let text = gather_text(&registry);
    assert!(text.contains("headgate_jobs_admitted_total{queue=\"critical\"} 3"));
    assert!(
        text.contains("headgate_jobs_rejected_total{policy=\"rate_class\",queue=\"critical\"} 2")
    );
    assert!(text.contains("headgate_jobs_completed_total{kind=\"mail.send\"} 1"));
    assert!(text.contains("headgate_worker_inflight{worker=\"worker-a\"} 3"));
    assert!(text.contains("headgate_worker_capacity{worker=\"worker-a\"} 8"));
    assert!(text.contains("headgate_worker_restart_requests_total{worker=\"worker-a\"} 1"));
    assert!(text.contains(
        "headgate_job_attempt_duration_seconds_sum{kind=\"mail.send\",outcome=\"success\",queue=\"critical\"} 0.25"
    ));
    assert!(!text.contains("must-not-be-a-label"));
}

#[test]
fn duplicate_and_partial_registration_leave_existing_registry_intact() {
    let registry = Registry::new();
    let first = headgate_prometheus::Telemetry::new(&registry).expect("first registration");
    first.on_event(Event::Quarantined {
        fingerprint: "not-a-label",
        crashes: 1,
    });
    assert!(headgate_prometheus::Telemetry::new(&registry).is_err());
    assert_eq!(
        gather_text(&registry)
            .matches("headgate_jobs_quarantined_total 1")
            .count(),
        1
    );
    assert!(!gather_text(&registry).contains("not-a-label"));

    let partial = Registry::new();
    let collision = Gauge::with_opts(Opts::new(
        "headgate_jobs_evicted_total",
        "An application-owned collector occupying a later Headgate name.",
    ))
    .expect("collision collector");
    collision.set(7.0);
    partial
        .register(Box::new(collision))
        .expect("register collision");
    assert!(headgate_prometheus::Telemetry::new(&partial).is_err());

    let text = gather_text(&partial);
    assert!(text.contains("headgate_jobs_evicted_total 7"));
    assert!(!text.contains("headgate_jobs_quarantined_total"));
}

fn gather_text(registry: &Registry) -> String {
    let mut bytes = Vec::new();
    TextEncoder::new()
        .encode(&registry.gather(), &mut bytes)
        .expect("encode metrics");
    String::from_utf8(bytes).expect("Prometheus text is UTF-8")
}
